lazyflow/operators/opRelabelConsecutive.py

```diff
@@ -1,26 +1,506 @@
+      for valid relabelings.
+    * `dtype` of the `Input` is checked (see `supportedDtypes`) and in case of
+      `uint16` changed to `uint32` to comply with vigra requirements. `uint16` is
+       what we have seen with cellpose data.
+    * `Input` is always reordered to `tzyxc`
+    """
+
     Input = InputSlot()
-    StartLabel = InputSlot(value=0)
-    Output = OutputSlot()
+    StartLabel = InputSlot(value=1)
+
+    CachedOutput = OutputSlot()
+    RelabelDict = OutputSlot()
+
+    CleanBlocks = OutputSlot()
+    CompressionEnabled = InputSlot(value=False)
+
+    SerializationInput = InputSlot(optional=True)
+    SerializationOutput = OutputSlot()
+
+    # total dummy slot, there because OpObjectExtraction expects this interface
+    BypassModeEnabled = InputSlot(value=False)
+
+    supportedDtypes = [numpy.uint8, numpy.uint16, numpy.uint32, numpy.uint64]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._opReorder5D = OpReorderAxes(parent=self, Input=self.Input, AxisOrder="tzyxc")
+        self._opDtypeConvert = OpPixelOperator(parent=self, Input=self._opReorder5D.Output, Function=lambda x: x)
+
+        self._opRelabelConsecutive = OpRelabelConsecutive5D(
+            parent=self,
+            Input=self._opDtypeConvert.Output,
+            StartLabel=self.StartLabel,
+            CompressionEnabled=self.CompressionEnabled,
+            SerializationInput=self.SerializationInput,
+        )
+        self._opReorderToInputOrder = OpReorderAxes(parent=self)
+        self.CachedOutput.connect(self._opReorderToInputOrder.Output)
+        self.RelabelDict.connect(self._opRelabelConsecutive.RelabelDict)
+        self.CleanBlocks.connect(self._opRelabelConsecutive.CleanBlocks)
+        self.SerializationOutput.connect(self._opRelabelConsecutive.SerializationOutput)
 
     def setupOutputs(self):
-        self.Output.meta.assignFrom(self.Input.meta)
+        tagged_input = self._opReorder5D.Output.meta.getTaggedShape()
+        tagged_blockshape = {k: v for k, v in tagged_input.items()}
+        tagged_blockshape["t"] = 1
+
+        input_dtype = self.Input.meta.dtype
+
+        if input_dtype == numpy.uint16:
+            self._opDtypeConvert.Function.setValue(lambda x: x.astype("uint32"))
+        else:
+            self._opDtypeConvert.Function.setValue(lambda x: x)
+
+        self._opReorderToInputOrder.AxisOrder.setValue("".join(self.Input.meta.getAxisKeys()))
+        self._opReorderToInputOrder.Input.connect(self._opRelabelConsecutive.Output)
+
+    def propagateDirty(self, slot, subindex, roi):
+        # all outputs connected directly
+        pass
+
+    def setInSlot(self, slot, subindex, roi, value):
+        assert slot == self.SerializationInput, f"Got unexpected slot {slot.name=}"
+        # setInSlot handled by downstream input slots
+        pass
+
+
+class OpRelabelConsecutive5D(OpUnblockedArrayCache):
```

Comment:
Not sure I understand what you mean by nesting here. You mean Having `OpRelabelConsecutive5D` as an object being used in `OpRelabelConsecutive`?

The latter operates in terms of input data dimensions, and `OpRelabelConsecutive5D` implementation is simplified by assuming 5D...

Reply:
Now I see OpRelabelConsecutive mirrors the internal architecture of OpLabelVolume pretty much 1:1, though it does read a lot cleaner here :)

This implicit relationship has tripped me up a few times during the review... I feel it could be helpful to have a noop OpObjectsABC or sth purely to document the connection, and/or rename `OpRelabelConsecutive` into `OpRelabelVolume`. I'm immediately also jumping to "well and extract the reorder-dtype-convert-MakeLabels-reorderToInput" cascade into the base class"... Or maybe OpRelabelConsecutive5D should really be a subclass of `OpLabelingABC`. But I remember you've mentioned going back and forth between implementations and discarding others, so I'm happy to accept what we have here was the cleanest way to solve this.

Reply:
it was good you left the comment as the common interface for object classification is pretty minimal as it turns out:

```
class OpLabelBase(Operator):
    """Minimal interface for Operators that takes an input volume and produces a
    consecutive labeling as an output volume"""

    Input = InputSlot()
    """Volume to (re)label, up to 5D"""

    CachedOutput = OutputSlot()
    """Cached label image - internally always whole time slices are stored
    Axistags and shape are the same as on the Input, dtype is an integer
    datatype.
    This slot extends the ROI to the full xyz volume (c and t are unaffected)
    and computes the labeling for the whole volume. As long as the input does
    not get dirty, subsequent requests to this slot guarantee consistent
    labelings.
    """

    CleanBlocks = OutputSlot()
    """Expose for Cache serialization, slicings for clean blocks stored in the cache"""
```

will probably do the job (still shifting things around a bit). It's a pity that not more of the interface can be changed, but well...